src/pip/_internal/network/lazy_wheel.py

```diff
@@ -131,15 +135,207 @@ def writable(self) -> bool:
         """Return False."""
         return False
 
-    def __enter__(self) -> "LazyZipOverHTTP":
+    def write(self, s: bytes) -> int:
+        raise NotImplementedError
+
+    def writelines(self, lines: Iterable[bytes]) -> None:
+        raise NotImplementedError
+
+    def __enter__(self) -> ReadOnlyIOWrapper:
         self._file.__enter__()
         return self
 
     def __exit__(self, *exc: Any) -> None:
         self._file.__exit__(*exc)
 
+    def __iter__(self) -> Iterator[bytes]:
+        raise NotImplementedError
+
+    def __next__(self) -> bytes:
+        raise NotImplementedError
+
+
+# The central directory for tensorflow_gpu-2.5.3-cp38-cp38-manylinux2010_x86_64.whl is
```

Comment:
What does “central directory” mean here?

Reply:
It's part of the structure of a zipfile (where a list of what's in it gets stored). Presumably this is saying that tensorflow has lots of files in it, so it's a good basis for knowing the max amount of data we need to read to be sure we grabbed the whole of the central directory. Or conversely, that grabbing 1MB is enough for even something as huge as tensorflow.

Reply:
I should probably expand the comment to mention there is a tradeoff here which we are optimizing by our choice of fetch granularity. The behavior we are seeing from PyPI (and presumably other hosts) described in #8670 is that each individual GET request for a large file (even if it's for a very small range) experiences a significant delay compared to requests for smaller files. We therefore want to minimize the number of requests we do for these very largest of files (e.g. `tensorflow-gpu==2.5.3`), but we can allow an extra request for very small files (since those don't experience the delay). That is what @McSinyx and @dholth's follow-up investigations identified and implemented, reproduced in this PR.

The "central directory" is the index which is always at the end of a zip file which describes its contents. It is parsed with the [`zipfile.ZipFile._RealGetContents()` method](https://github.com/python/cpython/blob/a47c13cae5b32e6f3d7532cc6dbb4e1ac31219de/Lib/zipfile/__init__.py#L1397). In order to do anything with the rest of the zip file (such as reading out some of its entries), we need to make sure this section is downloaded (see comments in `_prefetch_contiguous_dist_info()`).

By making this initial fetch size 1MB (and in particular: making this larger than the central directory record for the largest wheel I have been using to test against), we ensure that we will only make two GET requests total against PyPI for `tensorflow-gpu==2.5.3`. If there are wheels with larger central directories (this info is at the top of `zipinfo -v`), we should make sure the initial fetch size is large enough to cover those, too.

In fact, since the use case for this feature is mostly "internal corporate repos that don't want to or can't generate pypi-compatible indices", we might actually want to allow for the use case of "minimizing GET requests against wheels that aren't on PyPI, but are known to be extremely large". So now that I think about it, we might actually want to make the initial fetch size configurable, while the default should be set to "larger than the central directory record for any wheel on PyPI".

I think making the initial fetch size configurable would be a good idea for a followup PR which also turns on `fast-deps` by default. For the vast majority of use cases, I believe this default will be the most performant.

Reply:
Although if the metadata lookup caching from #12184 is merged, the performance of metadata fetching becomes much less important. So exposing the initial fetch size as an option might be overkill.

Reply:
> The behavior we are seeing from PyPI (and presumably other hosts) described in https://github.com/pypa/pip/issues/8670 is that each individual GET request for a large file (even if it's for a very small range) experiences a significant delay compared to requests for smaller files.

I can't seem to reproduce this from curl on the command line right now, but it is absolutely what I see when viewing the output of e.g. `pip install -vvv --use-feature=fast-deps`.

Reply:
Oh, sorry, an additional important point: **the `*.dist-info/` directory is also located at the end of the zip file for most wheels** (in fact, every wheel except for recent tensorflow uploads). This means that (unfortunately not for tensorflow), the choice of 1MB for the initial fetch size ensures we pull in (with just 1 single request) not just the central directory, but the contents of the `*.dist-info/` directory as well for most wheels, since those entries are located at the end of the zip file (this is what @dholth specifically discovered and implemented). That is also why we want to lean towards choosing a larger value for this initial fetch size, to pull in not just the central directory index but also the (probably compressed) metadata entries in the `*.dist-info/` directory.

Reply:
All of this complexity (including counting the number of requests performed for each wheel) is covered in test cases, so it shouldn't matter for reviewers.

Reply:
Ah OK, I know there’s a file index but didn’t know it’s called that. It makes sense now. (But the comment should be expanded to include some of the information above, or at least a link to the comment on GitHub.)

Reply:
> but didn’t know it’s called that

Yes, sorry, your question gave me a good excuse to write all that down is all. People say "central directory record" or "records"; just need to clarify that it's not the records from the inline file headers preceding each entry, which duplicate much (but not all) of the same information. This redundancy allows zip files to be processed in a streaming manner without parsing the central directory record first (see e.g. the rust `zip` crate's struct representing this limited form of inline metadata at https://docs.rs/zip/latest/zip/unstable/stream/struct.ZipStreamFileMetadata.html), which is a helpful optimization if your data is at the front of the file.

> But the comment should be expanded to include some of the information above

Yes, agreed. I think some of this should be covered in the `LazyWheelOverHTTP` docstring as well; I will add those docs along with the tests you suggested in your other comment.